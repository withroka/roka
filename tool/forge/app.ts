/**
 * Runtime support for forged applications.
 *
 * Provides functionality for applications compiled by `forge` at runtime.
 * Currently, only the {@linkcode version} function is available.
 *
 * @example
 * ```ts
 * import { version } from "@roka/forge/app";
 *
 * console.log(`App version: ${await version()}`);
 * ```
 *
 * @module
 */

import { PackageError, packageInfo } from "@roka/forge/package";
import { expandGlob } from "@std/fs";
import { dirname, join } from "@std/path";

/**
 * Returns the version of the current package.
 *
 * Useful for providing a version number to the user of a tool or application.
 *
 * The version is determined from whichever is available first:
 *  - release tags and {@link https://www.conventionalcommits.org | Conventional Commits} (local development)
 *  - config version from `deno.json` (deno run)
 *  - config version from `deno.json` in the dist directory (deno compile)
 *  - "(unknown)" if none of the above are available
 */
export async function version(): Promise<string> {
  return [
    `${await packageVersion()} (${Deno.build.target})`,
    `deno ${Deno.version.deno}`,
    `v8 ${Deno.version.v8}`,
    `typescript ${Deno.version.typescript}`,
  ].join("\n");
}

async function packageVersion(): Promise<string> {
  try {
    const pkg = await packageInfo();
    if (pkg.version) return pkg.version;
  } catch (e: unknown) {
    if (!(e instanceof PackageError)) throw e;
  }
  if (import.meta.dirname) {
    for await (
      const path of expandGlob("**/deno.json", {
        root: join(import.meta.dirname, "..", "..", "dist"),
        includeDirs: false,
      })
    ) {
      try {
        const pkg = await packageInfo({ directory: dirname(path.path) });
        if (pkg.version) return pkg.version;
      } catch (e: unknown) {
        if (!(e instanceof PackageError)) throw e;
      }
    }
  }
  return "(unknown)";
}
