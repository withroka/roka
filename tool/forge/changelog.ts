import type { Package } from "@roka/package";

/** Returns the changelog summary for a package since its last release. */
export function changelog(pkg: Package): string {
  return pkg.update?.changelog?.map((c) => ` * ${c.summary}`).join("\n") ?? "";
}
