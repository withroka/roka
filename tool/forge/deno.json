{
  "name": "@roka/forge",
  "version": "0.0.2",
  "exports": {
    ".": "./cli.ts",
    "./app": "./app.ts",
    "./bump": "./bump.ts",
    "./changelog": "./changelog.ts",
    "./compile": "./compile.ts",
    "./package": "./package.ts",
    "./release": "./release.ts"
  },
  "compile": {
    "main": "cli.ts",
    "permissions": {
      "env": ["GIT_NAME", "GIT_EMAIL", "GITHUB_TOKEN", "HOME"],
      "read": true,
      "write": true,
      "net": "api.github.com",
      "run": ["deno", "git"],
      "prompt": true
    }
  }
}
