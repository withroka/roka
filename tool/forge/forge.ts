// deno-lint-ignore-file no-console
/**
 * A monorepo tool for Deno workspaces.
 *
 * Roka ‚Äú**forge**‚Äù is crafted for the community for managing Deno packages
 * developed on GitHub. It can compile binaries, calculate versions, and even
 * create GitHub releases. It works best with monorepos of multiple packages
 * published to JSR.
 *
 * To get started, just run **forge** with `deno run -A jsr:@roka/forge` and
 * you‚Äôre all set! No need for any configuration.
 *
 * ## Packages
 *
 * The repository is treated as a list of Deno packages by **forge**. It
 * supports simple projects with a single package and monorepos that use
 * [Deno workspaces](https://docs.deno.com/runtime/fundamentals/workspaces/).
 *
 * ```sh
 * deno run -A jsr:@roka/forge list
 * ```
 * ```
 * üì¶ example @roka/example 1.2.3
 * üì¶ testing @roka/testing 0.1.0
 * ```
 *
 * By default, every **forge** command applies to all packages in the
 * repository. However, you can filter packages by specifying a list of package
 * directories or using glob patterns.
 *
 * ```sh
 * deno run -A jsr:@roka/forge list "exam*"
 * ```
 *
 * ## Versions
 *
 * Versioning in **forge** is only effective when the repository adheres to
 * {@link https://www.conventionalcommits.org | Conventional Commits}. Any bug
 * fix results in a patch version, while a new feature is introduced in a minor
 * version. Commits marked as _breaking changes_ trigger a new major version.
 *
 * A {@link https://semver.org | semantic version} is calculated for every
 * package at every commit. The new version begins with the latest release, or
 * _0.0.0_ if the package has no releases. It then tracks the commits made
 * since that release.
 *
 * ```sh
 * deno run -A jsr:@roka/forge changelog --emoji
 * ```
 * ```
 * üè∑Ô∏è example@2.0.0-pre.3+fedcba9
 *
 *   test(example): forgot unit-testing (#6)
 *   style(example): nicer code style (#5)
 *   fix(example)!: really fix bug with breaking change (#4)
 *   revert(example): revert bug fix, it broke something (#3)
 *   fix(example) fix bug (#2)
 *   feat(example): introduce new feature (#1)
 * ```
 *
 * The current package version will include a pre-release number, which is
 * calculated based on the number of commits that make a version change.
 * Additionally, the version will have a commit hash appended to it, which is
 * derived from the last commit that updated the package.
 *
 * _**WARNING**: Simple repositories with a single package still need to
 * specify a scope for every commit for that package. For example, a commit
 * summary like ‚Äúfeat: feature‚Äù does not automatically lead to a version
 * change. Instead, it should be formatted as ‚Äúfeat(example): feature‚Äù to
 * indicate that the feature was added to the package._
 *
 * ## Releases
 *
 * ### Bump the versions
 *
 * To start a new release, let‚Äôs first update the package versions in the
 * configuration files. We‚Äôre talking about the `version` field in `deno.json`.
 * If you change this version, it means a new release is coming, and **forge**
 * can help you with that.
 *
 * ```sh
 * deno run -A jsr:@roka/forge bump --release --pr
 * ```
 *
 * The `--release` flag drops the pre-release and build
 * information from the current version, resulting in a version like _2.0.0_.
 * If we had omitted this flag, the release would have been a pre-release.
 *
 * The pull request generated by the `--pr` flag will include changes to the
 * configuration and changelog files. Review and merge these changes, and the
 * release will be ready to roll at the merged commit.
 *
 * At this point, we are ready to publish the packages to [JSR](http://jsr.io).
 * However, before that, let‚Äôs proceed to create a release on GitHub, which is
 * the second step.
 *
 * ### Release on GitHub
 *
 * Great news! The version change is now merged, and GitHub knows the commit
 * hash. Let‚Äôs create a release right here!
 *
 * ```sh
 * deno run -A jsr:@roka/forge release example --draft
 * ```
 *
 * The draft release created with **forge** on GitHub will have the new version
 * number, the commit changelog, and a link to the package documentation on
 * JSR. If everything looks good, go ahead and publish the release. This will
 * create a new release tag for the released package, like _example@2.0.0_.
 *
 * ### Publish to JSR
 *
 * Finally, at this tagged commit, we can publish our packages to JSR.
 *
 * ```sh
 * deno publish
 * ```
 *
 * Simply running this command will work, because the `deno.json` file is
 * up-to-date, and all our changes are committed. Ideally, this step should be
 * executed using a workflow with GitHub actions.
 *
 * ### Automate with Actions
 *
 * Check out the workflows in the [roka](https://github.com/withroka/roka)
 * repository to see how we can automate all three steps: _bump_, _release_,
 * and _publish_. With **forge** taking care of most of the work, we can
 * either chill or find more time for coding. üíÜ‚Äç‚ôÄÔ∏è
 *
 * ## Assets
 *
 * _**WARNING**: This feature is highly experimental._
 *
 * The tool supports a non-standard `compile` extension in the `deno.json`
 * file. Any package with this field will generate release assets during a
 * release.
 *
 * ```json
 * {
 *   "name": "@roka/example",
 *   "version": "2.0.0",
 *   "compile": {
 *     "main": "example.ts"
 *   }
 * }
 * ```
 *
 * A package with the `compile` configuration will be compiled into a binary
 * for every supported Deno [target](https://docs.deno.com/go/compile). These
 * compiled binaries will be bundled and uploaded to the GitHub release as
 * assets.
 *
 * These assets can be created manually as well.
 *
 * ```sh
 * deno run -A jsr:@roka/forge compile example --bundle --checksum
 * ```
 * ```
 * üì¶ Compiled example
 *
 *   üè∫ dist/example/2.0.0/aarch64-apple-darwin.tar.gz
 *   üè∫ dist/example/2.0.0/sha256.txt
 * ```
 *
 * You can even install these binaries for the local user. The **forge** tool
 * itself can be compiled and installed from its source in this manner.
 *
 * ```sh
 * deno run -A jsr:@roka/forge compile forge --install
 * ```
 *
 * Now we can simply run **forge** by its name.
 *
 * ```sh
 * forge --version
 * ```
 * ```
 * forge 0.1.0 (aarch64-apple-darwin)
 * deno 2.2.3
 * v8 13.4.114.11-rusty
 * typescript 5.7.3
 * ```
 *
 * Magic! üîÆ
 *
 * ## Modules
 *
 * This library also offers programmatic functionality through the following
 * modules:
 *
 *  -  {@link [package]}: Retrieve package information.
 *  -  {@link [changelog]}: Generate changelogs.
 *  -  {@link [compile]}: Create binary executables.
 *  -  {@link [bump]}: Bump package versions.
 *  -  {@link [release]}: Create GitHub releases.
 *  -  {@link [version]}: Provide version from compiled binaries.
 *  -  {@link [testing]}: Write tests for **forge**.
 *
 * @todo Add documentation for GitHub workflows.
 * @todo Gracefully handle errors in the CLI.
 *
 * @module forge
 */

import { Command, EnumType, ValidationError } from "@cliffy/command";
import { Table } from "@cliffy/table";
import { pool, pooled } from "@roka/async/pool";
import { bump } from "@roka/forge/bump";
import { changelog } from "@roka/forge/changelog";
import { compile, targets } from "@roka/forge/compile";
import {
  commits,
  type Package,
  releases,
  workspace,
} from "@roka/forge/package";
import { release } from "@roka/forge/release";
import { version } from "@roka/forge/version";
import { join, relative } from "@std/path";

function listCommand() {
  return new Command()
    .description("List packages and versions.")
    .example("forge list", "List all packages.")
    .example("forge list --modules", "List all modules.")
    .arguments("[packages...:file]")
    .option("--modules", "Print exported package modules.", { default: false })
    .action(async ({ modules }, ...filters) => {
      const packages = await workspace({ filters });
      new Table().body(
        packages.map((pkg) => [
          "üì¶",
          pkg.directory,
          modules ? modulesText(pkg) : pkg.config.name,
          pkg.config.version !== undefined ? pkg.version : undefined,
          ...(pkg.latest?.version !== pkg.config.version)
            ? ["üö®", pkg.latest?.version, "üëâ", pkg.config.version]
            : [],
        ]),
      ).render();
    });
}

function modulesText(pkg: Package): string | undefined {
  const name = pkg.config.name;
  if (name === undefined) return undefined;
  const exports = pkg.config.exports ?? {};
  if (typeof exports === "string") return name;
  return Object.keys(exports)
    .map((key) => join(name, relative(".", key)))
    .join("\n");
}

function changelogCommand() {
  return new Command()
    .description("Generate changelogs.")
    .example("forge changelog", "List unreleased changes.")
    .example("forge changelog --type feat --no-breaking", "List new features.")
    .example("forge changelog --markdown --all", "All releases in Markdown.")
    .option("--all", "Generate changelog for all releases.")
    .option("--type=<type:string>", "Commit type.", { collect: true })
    .option("--breaking", "Only breaking changes.")
    .option("--no-breaking", "Skip breaking changes of filtered types.")
    .option("--emoji", "Use emoji for commit summaries.", { default: false })
    .option("--markdown", "Generate Markdown.", { default: false })
    .arguments("[packages...:file]")
    .action(async ({ all, type, breaking, emoji, markdown }, ...filters) => {
      const packages = await workspace({ filters });
      const options = {
        ...type && { type },
        ...breaking !== undefined && { breaking },
        ...markdown ? {} : { markdown: { heading: "üè∑Ô∏è  ", bullet: "  " } },
        emoji,
      };
      async function* changelogs(pkg: Package) {
        const log = await commits(pkg, {
          ...options,
          ...pkg.latest && { range: { from: pkg.latest.tag } },
        });
        if (log.length) {
          yield changelog(log, {
            ...options,
            title: `${pkg.name}@${pkg.version}`,
          });
        }
        if (!all) return;
        const releasesWithCommits = pooled(
          await releases(pkg),
          async (release) => ({
            ...release,
            commits: await commits(pkg, { ...options, ...release }),
          }),
        );
        for await (const release of releasesWithCommits) {
          if (release.commits.length) {
            yield changelog(release.commits, {
              ...options,
              title: `${pkg.name}@${release.tag.name}`,
            });
          }
        }
      }
      for (const pkg of packages) {
        for await (const log of changelogs(pkg)) {
          console.log(log.replace(/ ‚ôªÔ∏è /g, " ‚ôªÔ∏è  "));
        }
      }
    });
}

function compileCommand(targets: string[]) {
  return new Command()
    .description("Compile packages into binary executables.")
    .example("forge compile", "Compile packages.")
    .example("forge compile --install", "Install binaries.")
    .arguments("[packages...:file]")
    .type("target", new EnumType(targets))
    .option("--target=<architecture:target>", "Target OS architecture.", {
      collect: true,
    })
    .option("--bundle", "Zip and bundle artifacts.", { default: false })
    .option("--checksum", "Create a checksum file.", { default: false })
    .option("--install=[directory:file]", "Install for local user.")
    .option("--concurrency=<number:number>", "Max concurrent compilations.")
    .action(async (options, ...filters) => {
      const packages = (await workspace({ filters }))
        .filter((pkg) => pkg.config.compile);
      await pool(
        packages,
        async (pkg) => {
          const artifacts = await compile(pkg, options);
          console.log(`üì¶ Compiled ${pkg.name}`);
          console.log();
          artifacts.forEach((artifact) => console.log("  üè∫", artifact));
          if (options.install) console.log(`  üß© Installed ${pkg.name}`);
          console.log();
        },
        options,
      );
    });
}

function bumpCommand() {
  return new Command()
    .description("Bump versions on package config files.")
    .example("forge bump", "Bump versions.")
    .example("forge bump --release", "Bump to the next release version.")
    .example("forge bump --release --pr", "Create a version bump PR.")
    .example("forge bump --changelog=CHANGELOG.md", "Update changelog file.")
    .arguments("[packages...:file]")
    .option("--release", "Bump to the next release.", { default: false })
    .option("--changelog=<file:string>", "Update changelog file.")
    .option("--pr", "Create a pull request.", { default: false })
    .option("--emoji", "Use emoji for commit changelog.", { default: false })
    .env("GIT_NAME=<name:string>", "Git user name for the bump commit.", {
      prefix: "GIT_",
    })
    .env("GIT_EMAIL=<email:string>", "Git user e-mail for the bump commit.", {
      prefix: "GIT_",
    })
    .env(
      "GITHUB_TOKEN=<token:string>",
      "GitHub personal token for GitHub actions.",
      { prefix: "GITHUB_" },
    )
    .action(async (options, ...filters) => {
      const packages = (await workspace({ filters }))
        .filter((pkg) => pkg.config.version !== undefined);
      if (!packages.length) console.log("üì¶ No packages to release");
      const pr = await bump(packages, options);
      if (pr) {
        console.log(`üöÄ Created bump pull request`);
        console.log();
        console.log(`  [${pr.url}]`);
        console.log();
      } else {
        console.log("üì¶ Bumped package versions");
      }
    });
}

function releaseCommand() {
  return new Command()
    .description("Creates releases for updated packages.")
    .example("forge release", "Create releases and tags for all updates.")
    .example("forge release --draft", "Create draft releases for all updates.")
    .option("--draft", "Create a draft release.", { default: false })
    .option("--emoji", "Use emoji for commit summaries.", { default: false })
    .arguments("[packages...:file]")
    .env(
      "GITHUB_TOKEN=<token:string>",
      "GitHub personal token for GitHub actions.",
      { prefix: "GITHUB_", required: true },
    )
    .action(async (options, ...filters) => {
      const packages = (await workspace({ filters }))
        .filter((pkg) => pkg.config.version !== pkg.latest?.version);
      if (!packages.length) console.log("üì¶ No packages to release");
      await pool(packages, async (pkg) => {
        const [rls, assets] = await release(pkg, options);
        console.log(`üöÄ Created release ${pkg.name}@${pkg.version}`);
        console.log();
        console.log(`  [${rls.url}]`);
        console.log();
        if (assets.length) {
          assets.forEach((x) => console.log(`  üè∫ ${x.name}`));
          console.log();
        }
      }, { concurrency: 1 });
    });
}

if (import.meta.main) {
  let verbose = false;
  await new Command()
    .name("forge")
    .version(await version({ build: true, deno: true }))
    .description("Manage packages.")
    .example("forge", "List all packages.")
    .example("forge list 'core/*'", "List packages in the 'core' directory.")
    .example("forge compile --install", "Compile and install all binaries.")
    .example("forge bump --pr", "Bump versions and create a bump PR.")
    .example("forge release --draft", "Create releases with compiled assets.")
    .usage("<command> [options] [packages...]")
    .option("--verbose", "Print additional information.", {
      hidden: true,
      global: true,
      action: () => verbose = true,
    })
    .error((error, cmd) => {
      if (error instanceof ValidationError) {
        cmd.showHelp();
        console.error(`‚ùå ${error.message}`);
        Deno.exit(1);
      }
      const errors = (error instanceof AggregateError) ? error.errors : [error];
      if (error instanceof AggregateError) error = error.errors[0];
      for (const error of errors) {
        console.error(`‚ùå ${error.message}`);
        if (verbose) console.error(error);
        else if (error["cause"] && error["cause"]["error"]) {
          console.error(error.cause.error);
        }
      }
      Deno.exit(2);
    })
    .default("list")
    .command("list", listCommand())
    .command("changelog", changelogCommand())
    .command("compile", compileCommand(await targets()))
    .command("bump", bumpCommand())
    .command("release", releaseCommand())
    .parse();
}
