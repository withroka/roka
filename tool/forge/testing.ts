/**
 * This module provides utilities to create packages for testing.
 *
 * ```ts
 * import { tempPackage } from "@roka/forge/testing";
 * await using pkg = await tempPackage({ name: "@scope/name" });
 * ```
 *
 * @module testing
 */

import { type Config, type Package, packageInfo } from "@roka/forge/package";
import { join } from "@std/path";

/**
 * Creates a test package under given directory with given configuration.
 *
 * @example
 *
 * ```ts
 * import { testPackage } from "@roka/forge/testing";
 * import { tempDirectory } from "@roka/testing/temp";
 * import { assertEquals } from "@std/assert";
 *
 * await using directory = await tempDirectory();
 * const pkg = await testPackage(directory.path(), {
 *   name: "@scope/name",
 *   version: "1.2.3",
 * });
 *
 * assertEquals(pkg.name, "name");
 * assertEquals(pkg.version, "1.2.3");
 * ```
 */
export async function testPackage(
  directory: string,
  config: Config & { root?: string },
): Promise<Package> {
  await Deno.mkdir(directory, { recursive: true });
  await Deno.writeTextFile(
    join(directory, "deno.json"),
    JSON.stringify(config, undefined, 2),
  );
  const pkg = await packageInfo({ directory });
  if (config.root) pkg.root = config.root;
  return pkg;
}

/**
 * Creates a test package with given configuration.
 *
 * The package will be created under a temporary directory, which will be
 * automatically removed when the package is disposed.
 *
 * @example
 *
 * ```ts
 * import { tempPackage } from "@roka/forge/testing";
 * import { assertEquals } from "@std/assert";
 *
 * await using pkg = await tempPackage({
 *   name: "@scope/name",
 *   version: "1.2.3",
 * });
 *
 * assertEquals(pkg.name, "name");
 * assertEquals(pkg.version, "1.2.3");
 * ```
 */
export async function tempPackage(
  config: Config,
): Promise<AsyncDisposable & Package> {
  const directory = await Deno.makeTempDir();
  const pkg = await testPackage(directory, config);
  return Object.assign(pkg, {
    [Symbol.asyncDispose]: () => Deno.remove(directory, { recursive: true }),
  });
}
