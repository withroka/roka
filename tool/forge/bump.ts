/**
 * This module provides the {@linkcode bump} function, which updates the
 * version string in package configuration file (`deno.json`). The version is
 * calculated based on the
 * {@link https://www.conventionalcommits.org | Conventional Commits} since
 * the last release and {@link https://semver.org | semantic versioning}.
 *
 * If the {@linkcode BumpOptions.release | release} option is set, the next
 * release version is written to the package configuration. Otherwise, the
 * current pre-release version is written.
 *
 * If the {@linkcode BumpOptions.pr | pr} option is set, a pull request is
 * created with the updated version information on GitHub. Once this pull
 * request is merged, a release can be created using the {@linkcode [release]}
 * module.
 *
 * ```ts
 * import { bump } from "@roka/forge/bump";
 * import { workspace } from "@roka/forge/package";
 * async function usage() {
 *   const packages = await workspace();
 *   await bump(packages, { pr: true });
 * }
 * ```
 *
 * @module bump
 */

import { pool } from "@roka/async/pool";
import { changelog } from "@roka/forge/changelog";
import { type Package, PackageError } from "@roka/forge/package";
import { github, type PullRequest, type Repository } from "@roka/github";
import { assertExists } from "@std/assert";
import { common, join } from "@std/path";
import { format, parse } from "@std/semver";

const BUMP_BRANCH = "automated/bump";

/** Options for the {@linkcode bump} function. */
export interface BumpOptions {
  /** GitHub access token. */
  token?: string;
  /**
   * GitHub repository to use.
   *
   * If not defined, the repository is determined from the package directories.
   */
  repo?: Repository;
  /** Git user for the bump commit. */
  user?: {
    /** Name of the user. */
    name?: string;
    /** Email of the user. */
    email?: string;
  };
  /** Bump to the next release version, instead of a pre-release of it. */
  release?: boolean;
  /**
   * Update given file with the generated changelog.
   *
   * If a relative path is provided, the path will be resolved from the current
   * working directory.
   */
  changelog?: string;
  /** Create a pull request. */
  pr?: boolean;
  /** Use emoji in commit summaries. */
  emoji?: boolean;
}

/**
 * Updates the version numbers on package configuration files (`deno.json`).
 *
 * The version for the package is calculated using the latest release tag and
 * the {@link https://www.conventionalcommits.org | Conventional Commits} for
 * the package since that release. If the changelog is not empty, the version
 * will be a pre-release version. If {@linkcode BumpOptions.release | release}
 * is set, the version of the next release will be written, dropping prerelase
 * and build information from the version string.
 *
 * When working with pull requests, if there is an open PR, it will be updated
 * with the new version information.
 *
 * @param pkg Package to bump.
 * @throws {PackageError} If the package does not have an update.
 *
 * @todo Recalculate versions when the bump PR is rebased.
 */
export async function bump(
  packages: Package[],
  options?: BumpOptions,
): Promise<PullRequest | undefined> {
  const bumps = packages
    .map((pkg) => ({
      pkg,
      bump: format({
        ...parse(pkg.version),
        ...options?.release ? { prerelease: [], build: [] } : {},
      }),
    }))
    .filter(({ pkg, bump }) => pkg.config.version !== bump);
  packages = await pool(bumps, ({ pkg, bump }) => updateConfig(pkg, bump));
  if (packages.length === 0) throw new PackageError("No packages to bump");
  if (options?.changelog) await updateChangelog(packages, options);
  if (!options?.pr) return undefined;
  return createPullRequest(packages, options);
}

async function updateConfig(pkg: Package, version: string) {
  const config = { ...pkg.config, version };
  await Deno.writeTextFile(
    join(pkg.directory, "deno.json"),
    JSON.stringify(config, undefined, 2) + "\n",
  );
  pkg.version = version;
  pkg.config.version = version;
  return pkg;
}

async function updateChangelog(packages: Package[], options?: BumpOptions) {
  assertExists(options?.changelog, "Changelog file was not passed");
  const prepend = packages.map((pkg) =>
    changelog(pkg.changes ?? [], {
      ...options,
      title: `${pkg.name}@${pkg.version}`,
    })
  ).join("\n");
  let existing = "";
  try {
    existing = await Deno.readTextFile(options?.changelog);
  } catch (e: unknown) {
    if (!(e instanceof Deno.errors.NotFound)) throw e;
  }
  await Deno.writeTextFile(
    options?.changelog,
    [prepend, ...existing && [existing]].join("\n"),
  );
}

async function createPullRequest(
  packages: Package[],
  options?: BumpOptions,
): Promise<PullRequest> {
  if (packages.length === 0) throw new PackageError("No packages to bump");
  const directory = common(packages.map((pkg) => pkg.directory));
  const { repo = await github(options).repos.get({ directory }), user } =
    options ?? {};
  const title = packages.length === 1
    ? `chore: bump ${packages[0]?.name} version`
    : "chore: bump versions";
  const body = packages.map((pkg) =>
    changelog(pkg.changes ?? [], {
      ...options,
      title: `${pkg.name}@${pkg.version}`,
    })
  ).join("\n");
  await repo.git.branches.checkout({ new: BUMP_BRANCH });
  await repo.git.config.set({ ...user && { user } });
  await repo.git.index.add([
    ...packages.map((pkg) => join(pkg.directory, "deno.json")),
    ...options?.changelog ? [options?.changelog] : [],
  ]);
  await repo.git.commits.create(title, { body });
  let [pr] = await repo.pulls.list({ title, closed: false });
  if (pr) {
    await repo.git.commits.push({ force: true, branch: BUMP_BRANCH });
    pr.update({ body });
  } else {
    await repo.git.commits.push({ branch: BUMP_BRANCH });
    pr = await repo.pulls.create({ title, body, draft: true });
  }
  return pr;
}
