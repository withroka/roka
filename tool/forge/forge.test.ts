import { git } from "@roka/git";
import { tempRepository } from "@roka/git/testing";
import { fakeRepository } from "@roka/github/testing";
import { fakeArgs, fakeConsole, fakeEnv } from "@roka/testing/fake";
import { assertExists } from "@std/assert";
import { common, join } from "@std/path";
import { assertSnapshot } from "@std/testing/snapshot";
import { forge } from "./forge.ts";
import { tempWorkspace } from "./testing.ts";

const WORKSPACE = {
  configs: [
    {
      name: "@scope/name1",
      version: "1.0.0",
      exports: { ".": "./name2.ts", "./main": "./main.ts" },
      forge: { main: "main.ts", target: ["aarch64-unknown-linux-gnu"] },
    },
    {
      name: "@scope/name2",
      version: "2.0.0",
      exports: "./name1",
    },
    {
      name: "@scope/name3",
      version: "3.0.0",
    },
    {
      name: "@scope/name4",
      version: "4.0.0",
    },
    {
      version: "0.0.0",
      name: "@scope/dir/name5",
    },
    {
      name: "@scope/dir/name6",
    },
  ],
  commits: [
    { summary: "initial", tags: ["name1@1.0.0-pre.1", "name4@4.0.0"] },
    { summary: "feat(name2): name2", tags: ["name2@1.0.0"] },
    { summary: "fix(name1): bug", tags: ["name2@2.0.0"] },
    { summary: "refactor(name2): rewrite" },
    { summary: "feat(name2): feature", tags: ["name3@2.0.0"] },
    { summary: "docs(name3): fix typo" },
    { summary: "refactor(name4)!: redesign api" },
    { summary: "style(name5): tabs over spaces" },
  ],
};

async function run(context: Deno.TestContext) {
  await using remote = await tempRepository();
  await using packages = await tempWorkspace({
    ...WORKSPACE,
    repo: { clone: remote.path() },
  });
  assertExists(packages[0]);
  const pkg = packages[0];
  const root = common(packages.map((pkg) => pkg.root));
  const repo = fakeRepository({ url: "<url>", git: git({ cwd: root }) });
  await Deno.writeTextFile(
    join(pkg.directory, "main.ts"),
    "console.log('Hello, World!');",
  );
  using _args = fakeArgs(
    context.name
      .replaceAll("[package]", pkg.name)
      .replaceAll("[packages...]", packages.map((pkg) => pkg.name).join(" "))
      .replaceAll("[pattern]", "name*")
      .replaceAll("[directory]", root)
      .replaceAll("<file>", join(root, "CHANGELOG.md"))
      .replaceAll("<type>", "feat")
      .replaceAll("<target>", "aarch64-unknown-linux-gnu")
      .split(" ").slice(1),
  );
  using _env = fakeEnv({ GITHUB_TOKEN: "token" });
  using console = fakeConsole();
  const cwd = Deno.cwd();
  try {
    Deno.chdir(root);
    const code = await forge({ repo });
    return {
      code,
      // deno-lint-ignore no-console
      output: console
        .output({ stripAnsi: true, stripCss: true, trimEnd: true, wrap: "\n" })
        .replace(/(?<=\n)((?:.*?):\s*)\d+(\.\d+)+(?:.*)?/g, "$1<version>")
        .replace(/(?<=(\d+\.\d+\.\d+-\w+\.\d+)\+)(.......)/g, "<hash>")
        .replaceAll(root, "<directory>"),
    };
  } finally {
    Deno.chdir(cwd);
  }
}

async function test(t: Deno.TestContext) {
  await assertSnapshot(t, await run(t));
}

Deno.test("forge --help", test);
Deno.test("forge list", test);
Deno.test("forge list --modules [packages...]", test);
Deno.test("forge list [pattern]", test);
Deno.test("forge list [directory]", test);
Deno.test("forge changelog", test);
Deno.test("forge changelog --all --emoji", test);
Deno.test("forge changelog --type <type>", test);
Deno.test("forge changelog --type <type> --no-breaking --all", test);
Deno.test("forge changelog --breaking --markdown", test);
Deno.test("forge compile --target <target> --bundle --install", test);
Deno.test("forge bump --release --pr --changelog=<file> --emoji", test);
Deno.test("forge release --draft --emoji", test);
