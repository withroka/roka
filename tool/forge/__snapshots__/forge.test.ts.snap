export const snapshot = {};

snapshot[`forge list 1`] = `
"
🚨 @scope/name1     0.0.0 👉 1.0.0
📦 @scope/name2     2.1.0-pre.1+<hash>
🚨 @scope/name3     2.0.0 👉 3.0.0
📦 @scope/name4     4.0.0
📦 @scope/dir/name5
"
`;

snapshot[`forge list --modules name1 name2 1`] = `
"
🚨 @scope/name1 0.0.0 👉 1.0.0

  🧩 (default)  <root>/name1/name2.ts
  🧩 main       <root>/name1/main.ts

📦 @scope/name2 2.1.0-pre.1+<hash>

  🧩 (default)  <root>/name2/name1

"
`;

snapshot[`forge list name* 1`] = `
"
🚨 @scope/name1     0.0.0 👉 1.0.0
📦 @scope/name2     2.1.0-pre.1+<hash>
🚨 @scope/name3     2.0.0 👉 3.0.0
📦 @scope/name4     4.0.0
📦 @scope/dir/name5
"
`;

snapshot[`forge list dir/* 1`] = `
"
📦 @scope/dir/name5
"
`;

snapshot[`forge changelog 1`] = `
"
🏷️  name1@1.0.0

  fix(name1): bug

🏷️  name2@2.1.0-pre.1+<hash>

  feat(name2): feature
  refactor(name2): rewrite

🏷️  name3@3.0.0

  style(name3)!: tabs over spaces

🏷️  name4@4.0.0

  docs(name4): fix typo

"
`;

snapshot[`forge changelog --all --emoji 1`] = `
"
🏷️  name1@1.0.0

  🐛 bug

🏷️  name2@2.1.0-pre.1+<hash>

  ✨ feature
  ♻️ rewrite

🏷️  name3@3.0.0

  🎨 tabs over spaces 💥

🏷️  name4@4.0.0

  📝 fix typo

"
`;

snapshot[`forge changelog --type docs --no-breaking 1`] = `
"
🏷️  name4@4.0.0

  docs(name4): fix typo

"
`;

snapshot[`forge changelog --breaking --markdown 1`] = `
"
## name3@3.0.0

- style(name3)!: tabs over spaces

"
`;

snapshot[`forge compile --target aarch64-unknown-linux-gnu --bundle --install=<root>/install --concurrency=2 1`] = `
"
📦 Compiled name1

  🏺 <root>/dist/name1/1.0.0/aarch64-unknown-linux-gnu.tar.gz
  💾 Installed name1

"
`;

snapshot[`forge bump --release --pr --changelog=<root>/CHANGELOG.md --emoji 1`] = `
"
🚀 Created bump pull request

  [<url>/pulls/1]

"
`;

snapshot[`forge release --draft --emoji 1`] = `
"
🚀 Created release name1@1.0.0

  [<url>/releases/name1@1.0.0]

  🏺 aarch64-unknown-linux-gnu.tar.gz
  🏺 sha256.txt

🚀 Created release name3@3.0.0

  [<url>/releases/name1@1.0.0]

"
`;
